modeling/local-slice-algorithm.go

```diff
@@ -0,0 +1,132 @@
+	// endpointsAvailable stores 'endpointDeviation' -- zones with 'extra'
+	// endpoints
+	endpointsAvailable := list.New()
+	// endpointsNeeded stores zones with number of endpoints needed
+	endpointsNeeded := map[string]int{}
+	for zoneName, zone := range region.zoneDetails {
+		var localGroup EndpointSliceGroup
+		localGroup.Label = zoneName
+		// this local sliceGroup should only receive traffic from current zone,
+		// this map tracks weights of traffic from different zones to this
+		// sliceGroup
+		localGroup.Weights = map[string]float64{zoneName: 1.0}
+		// this map stores composition of this sliceGroup, zoneName - number of
+		// endpoints from zoneName
+		localGroup.Composition = map[string]int{}
+
+		// the expected number of endpoints is with the same ratio as the number
+		// of nodes in this zone
+		expectedEndpoints := zone.nodesRatio * float64(region.totalEndpoints)
+		// deviation: - means need more endpoints from other sliceGroups, +
+		// means need give out endpoints to other sliceGroups
+		deviation := math.Round(float64(zone.Endpoints) - expectedEndpoints)
+		if deviation > 0 {
+			endpointsAvailable.PushBack(endpointDeviation{name: zoneName, deviation: int(deviation)})
+			localGroup.Composition[zoneName] = int(expectedEndpoints)
+		} else {
+			endpointsNeeded[zoneName] = int(-deviation)
+			localGroup.Composition[zoneName] = zone.Endpoints
+		}
+
+		sliceGroups[zoneName] = localGroup
+	}
+
+	err := balanceSliceGroups(endpointsAvailable, endpointsNeeded, sliceGroups)
+	if err != nil {
+		return nil, err
+	}
+	return sliceGroups, nil
+}
+
+// balanceSliceGroups helps to 'rent' endpoints from zones have extra endpoints
+// to EndpointSliceGroups that have insufficient ones
+func balanceSliceGroups(endpointsAvailable *list.List, endpointsNeeded map[string]int, sliceGroups map[string]EndpointSliceGroup) error {
+	for receiveZone, deviation := range endpointsNeeded {
+		for {
+			// the available stack is empty while there are still endpoints in
+			// need causes an error. This can happen when the approximation on
+			// deviation (calculated above) ends up in asymmetric sums of
+			// endpoints available and endpoints in need (in need > available)
+			if endpointsAvailable.Back() == nil {
+				return fmt.Errorf("endpoints available are less than needed")
+			}
+			sendZone, ok := endpointsAvailable.Back().Value.(endpointDeviation)
+			endpointsAvailable.Remove(endpointsAvailable.Back())
+			if !ok {
+				return fmt.Errorf("unexpected type error while asserting type to endpointDeviation")
+			}
+			if sendZone.deviation == deviation {
+				sliceGroups[receiveZone].Composition[sendZone.name] = deviation
+				break
+			}
+			if sendZone.deviation > deviation {
+				// there are extra endpoints remaining, put them back in the
+				// available stack, wait for the next customer
+				sliceGroups[receiveZone].Composition[sendZone.name] = deviation
+				sendZone.deviation -= deviation
+				endpointsAvailable.PushBack(sendZone)
```

Comment:
Related to comment above, I don't think we should remove and push back.

Reply:
Gonna change this list structure to map instead, I don't think its allowed to modify the value in the list directly. So I pop it out everytime.